### PR TITLE
Add a summary table into secret manager

### PR DESCRIPTION
**Pull request by premun**
https://github.com/dotnet/dnceng/issues/309

### CODE REVIEW

**Review by oleksandr-didyk — DISMISSED** [review]

File: src/SecretManager/Microsoft.DncEng.SecretManager.Tests/SynchronizeCommandTests.cs [hunk 1]
```diff
@@ -101,13 +99,16 @@ public class SynchronizeCommandTests
...
+        await TestCommand(
+            now,
+            """
+            storageLocation:
+              type: test
+            secrets:
+              expired:
+                type: test-secret
+            """,
+            "test",
```
Comment (dougbu) [inline src/SecretManager/Microsoft.DncEng.SecretManager.Tests/SynchronizeCommandTests.cs:111]: the minor formatting and whitespace changes in this PR make it _much_ more difficult to see the meat of the work. please do anything more than the core changes and perhaps trailing whitespace trimming (which is easy to ignore in GitHub) in separate PRs

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs [hunk 2]
```diff
@@ -52,7 +53,7 @@ public override async Task RunAsync(CancellationToken cancellationToken)
     {
         try
         {
-            _console.WriteLine($"Synchronizing secrets contained in {_manifestFile}");
+            _console.WriteLine($"📄 Synchronizing secrets contained in {_manifestFile}");
```
Comment (dougbu) [inline src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:56]: what's the new character here❔ I can't read what it's trying to indicate. was its addition intentional❔

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs [hunk 3]
```diff
@@ -63,6 +64,13 @@ public override async Task RunAsync(CancellationToken cancellationToken)
                 }
             }
 
+            Console.OutputEncoding = System.Text.Encoding.UTF8;
+            var problems = new List<(string, string)>();
```
Comment (dougbu) [inline src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:68]: nit: unnamed tuples make code less readable
```suggestion
            var problems = new List<(string name, string problem)>();
```

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs [hunk 4]
```diff
@@ -84,12 +92,15 @@ public override async Task RunAsync(CancellationToken cancellationToken)
 
             foreach (var (name, secret, secretType, secretReferences) in orderedSecretTypes)
             {
-                _console.WriteLine($"Synchronizing secret {name}, type {secret.Type}");
+                 _console.WriteLine($"Synchronizing secret {name}, type {secret.Type}");
+                var summaryStatus = (Status: string.Empty, Icon: string.Empty);
```
Comment (dougbu) [inline src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:96]: nit: again, please name the tuple values

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs [hunk 5]
```diff
@@ -213,8 +250,22 @@ public override async Task RunAsync(CancellationToken cancellationToken)
                 foreach (var (name, value) in existingSecrets)
                 {
                     _console.LogWarning($"Extra secret '{name}' consider deleting it.");
+                    table.AddRow(name, string.Empty, "⚠️", "Extra secret");
```
Comment (dougbu) [inline src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:253]: you don't have the secret type available here❔

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs [hunk 6]
```diff
@@ -213,8 +250,22 @@ public override async Task RunAsync(CancellationToken cancellationToken)
...
                 }
             }
+
+            _console.WriteLine(Environment.NewLine + Environment.NewLine + "🔎 Summary:" + Environment.NewLine);
+            _console.WriteLine(table.ToMinimalString());
+            _console.WriteLine("🔥 Detected additional problems:");
```
Comment (dougbu) [inline src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:259]: suggest checking whether `problems` is empty. skip this output if so

**Review by dougbu — CHANGES_REQUESTED** [review]

File: src/SecretManager/Microsoft.DncEng.SecretManager.Tests/SynchronizeCommandTests.cs (hunk 1 above)
Comment (premun) [reply src/SecretManager/Microsoft.DncEng.SecretManager.Tests/SynchronizeCommandTests.cs:111]: Whitespace changes can be turned off here:

image

Does that work for you?

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs (hunk 2 above)
Comment (premun) [reply src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:56]: I added this as a visual queue that a new keyvault is started to be processed.

One of the problems I have with the logs is to figure out which keyvault the problematic secret is from so I am looking for this log line a lot.

File: src/SecretManager/Microsoft.DncEng.SecretManager.Tests/SynchronizeCommandTests.cs (hunk 1 above)
Comment (premun) [reply src/SecretManager/Microsoft.DncEng.SecretManager.Tests/SynchronizeCommandTests.cs:111]: All whitespace (not just trailing) can be actually turned off. The tests were quite hard to read for me so I took the liberty but I can revert this file

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs (hunk 4 above)
Comment (premun) [reply src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:96]: These already have a name

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs (hunk 5 above)
Comment (premun) [reply src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:253]: Just name, the expires-on tag and possibly other tags.

File: Directory.Packages.props [hunk 7]
```diff
@@ -20,6 +20,7 @@
     <PackageVersion Include="Azure.Security.KeyVault.Secrets" Version="4.4.0" />
     <PackageVersion Include="Azure.Storage.Blobs" Version="12.13.1" />
     <PackageVersion Include="Azure.Storage.Queues" Version="12.11.0" />
+    <PackageVersion Include="ConsoleTables" Version="2.4.2" />
```
[the remark this answers is not in the document]
Comment (premun) [reply Directory.Packages.props:23]: Yes we do, but I checked/know this exact one before using it. The source code is public, the owner is a guy from the JetBrains Rider I know and the source code is a single `.cs` file:
https://github.com/khalidabuhakmeh/ConsoleTables/blob/master/src/ConsoleTables/ConsoleTable.cs

[the remark this answers is not in the document]
Comment (dougbu) [reply Directory.Packages.props:23]: alright. do we have a process for confirming use of 3rd party packages within the group❔ product teams tend to require an escalation before taking new external dependencies.

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs [hunk 8]
```diff
@@ -137,12 +152,16 @@ public override async Task RunAsync(CancellationToken cancellationToken)
                             }
                             else
                             {
-                                _console.Write($"For secret {name}, could not parse the {AzureKeyVault.NextRotationOnTag} tag with value {nextRotationOn}");
+                                var message = $"Missing {AzureKeyVault.NextRotationOnTag} tag, using the end of time as value";
```
Comment (dougbu) [inline src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:155]: Message is now incorrect. Should be reporting the problem parsing the value, ideally logging `nextRotationOn` as it did before.

**Review by dougbu — APPROVED** [review]

**Review by dougbu — CHANGES_REQUESTED** [review]
oops

File: src/SecretManager/Microsoft.DncEng.SecretManager.Tests/SynchronizeCommandTests.cs (hunk 1 above)
Comment (dougbu) [reply src/SecretManager/Microsoft.DncEng.SecretManager.Tests/SynchronizeCommandTests.cs]: I'm fully aware of that feature. it doesn't cover rewrapping lines or going from no indentation to some. thanks for reverting changes here

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs (hunk 2 above)
Comment (dougbu) [reply src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:56]: my point was mostly that the character is hard to discern and meant nothing to me. if it means something to you 🆗 but you may want to find something else

File: Directory.Packages.props (hunk 7 above)
[the remark this answers is not in the document]
Comment (premun) [reply Directory.Packages.props:23]: No, I don't think we don't but internal tooling is in a different boat than code that ends up in the product?

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs (hunk 2 above)
Comment (premun) [reply src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs:56]: Ah, yeah, no I can look for something else. The message is about opening a manifest file, so I put there a file emoji

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs (hunk 8 above)
Comment (premun) [reply src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs]: Ah, good catch

File: src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs (hunk 2 above)
Comment (premun) [reply src/SecretManager/Microsoft.DncEng.SecretManager/Commands/SynchronizeCommand.cs]: I put the one for "synchronization" 🔁 same as is used when a secret gets rotated. Should be more visible.

File: Directory.Packages.props (hunk 7 above)
[the remark this answers is not in the document]
Comment (missymessa) [reply Directory.Packages.props:23]: I don't have an issue with using this package, and I would imagine that we'd get CG warnings if there were issues with it. 

/cc @garath and @riarenas for opinions. 

@dougbu if you think we should have a formalized process for taking in new 3rd party dependencies, would you mind driving that process?

**Review by missymessa — APPROVED** [review]
LGTM shipit

**Review by dougbu — APPROVED** [review]
I should have updated my review after you fixed the incorrect message @premun. We can discuss 3rd party dependencies outside this PR&hellip;

**Comment by dougbu** [comment]
Now GitHub is doing the "last reviewer" dance 😦 I can't approve this and bet your vote was reset @missymessa

**Review by dougbu — APPROVED** [review]

**Comment by dougbu** [comment]
> I can't approve this

Weird, GitHub let me approve after telling me someone other than the last committer must approve. I'll enable auto-merge.

@AlitzelMendez any idea why there are two `Build Analysis` checks running on this PR❔

**Comment by missymessa** [comment]
> > I can't approve this
> 
> Weird, GitHub let me approve after telling me someone other than the last committer must approve. I'll enable auto-merge.
> 
> @AlitzelMendez any idea why there are two `Build Analysis` checks running on this PR❔

One is on staging and the other is on prod. We use our own repos to dogfood staging.

**Comment by dougbu** [comment]
> > @AlitzelMendez any idea why there are two Build Analysis checks running on this PR❔

> One is on staging and the other is on prod. We use our own repos to dogfood staging.

Ah, thanx for the explanation.

----

@Should we enable auto-merge to avoid further merge-from-main requirements❔

**Comment by missymessa** [comment]
I don't think that auto-merge is preferable. I ended up turning that off the Maestro PRs, even.

**Pull request merged**